### DISCUSSION TITLE
Automate events such as Peak Streak

### DISCUSSION THREAD

**Discussion by hb-robo** [RetroAchievements/RAWeb#2309] [Ideas]
Personally I think the biggest barrier to event involvement (aside from front page visibility for non-AOTW events) is that you have to participate through a forum post or Discord channel. I believe some events are good candidates to have their processes run automatically through the database and site pages. Instantaneous enrollment/participation would strengthen the feedback loop of the user and their progress, and eliminate the need for Event admins to respond to messages, assign rolls, and track player progress manually in off-site spreadsheets.

The event that seems easiest to automate is Peak Streak, where the player must earn one mastery per day for as long as they can manage.

For posterity here are the rules of the event:
> 1. Your streak begins the first time you post in the thread. Streaks earned prior to this date are irrelevant. 
> 2. Each day starts at 00:00 and ends at 23:59 UTC (server time).
> 3. "Mastering" a set means earning all achievements in Hardcore Mode. The time of mastery is based upon the time at which you earn the final achievement; when you start the set is irrelevant.
> 4. Mastering more than one game per day is allowed, though there is no advantage to doing so.
> 5. You may not delete achievements in a set that you have previously mastered to “remaster” the set. Similarly, you may not “remaster” a set following a set revision.
> 6. Usage of an alt account is not allowed.
> 7. Event badges and games with fewer than six achievements do not contribute toward streaks.
> 8. In the event that you require a manual unlock in order to master a set, please bring this to my attention and I will take into account the time at which you submitted the manual unlock request.

Right off the bat, three rules need consideration in an automated context:
* Rule 1 would be DOA, as if this system goes up all current streaks would have to be preserved, and therefore existing unlogged streaks would also be valid.
* Rule 5 necessitates a way to refer to earliest recorded mastery date, even if an achievement is reset. I am not sure such a feature of the database exists, but there is at least a way to tell if a set was mastered prior to a revision, so potentially this can be expanded on.
* Rule 8 would require a way to manually unlock an achievement such that the unlock timestamp is retroactive to the datetime of the unlock request. This may exist, I'm not sure.

**Comment by drisc** [comment]
The on-site events system proposal that I'm currently working on deals with event automation by checking enrolled users against a list of requirements which will automate tracking for the event as long as it's set up correctly in the event management system. That management system is one of the most complicated parts of the proposal since it needs to be flexible enough to allow for a variety of conditons to be tracked and will likely require updates over time as new event types are created.

For events that require daily posts for validation there would have to be a period of transfering the data into the events system somehow, this could be a manual process performed once to get everything input correctly or something automated at the DB level if a query can be crafted to check the relevent forum threads for the information that the event requires. The events team also keeps other records for events that would make data import smoother.

As you said with Rule 5 we can still see masteries prior to a revison since that record is not updated unless the user resets all their achievements and it still would appear on their wall.

Things like Rule 8 are manually tracked currently by the events team in documents so there would have to be a modifcation to the manual unlock system to flag that the achievement is elegable to be counted for the events of enrolled user.

The initial testbed for the events system will probably be AotW/AotM since the requirements requirements are so simple, along with LotM.

**Reply by hb-robo** [reply]
That's great! I have also thought about a similar page to the achievement set pages for AOTW/AOTM with target goals (target achievements, target leaderboard scores, certain progress criteria/status), could be repurposed for "quests" down the line too.

Sorry if I'm misunderstanding but is forum posting still going to be required for Peak Streak? Seems vestigial if this goes the way you're describing

**Reply by drisc** [reply]
For events like Peak Streak, once converted to the events system, the information would be parsed and handled by the events system with something like `if date = prev_date + 1 then peak_count + 1 else peak_count = 0` and that would be displayed on the users event information dashboard. So if they missed a day then the streak would drop to 0 and they would have to start over. Since the system is keeping track of all of this the forum post would no longer be required unless the event host wants it for data validation purposes.
